pkg/daemon/apply.go

```diff
@@ -0,0 +1,177 @@
+}
+
+func (a SchedulablePostAction)Execute(dn *Daemon, newConfig *mcfgv1.MachineConfig) error {
+	return drain.RunCordonOrUncordon(dn.drainer, dn.node, false)
+}
+
+func getFileNames(files []igntypes.File) []interface{} {
+	names := make([]interface{}, len(files))
+	for i, file := range files {
+		names[i] = file.Path
+	}
+	return names
+}
+
+func filesToMap(files []igntypes.File) map[string]igntypes.File {
+	fileMap := make(map[string]igntypes.File, len(files))
+	for _, file := range files {
+		fileMap[file.Path] = file
+	}
+	return fileMap
+}
+
+func getFileChanges(oldFilesConfig, newFilesConfig []igntypes.File) ([]igntypes.File, []igntypes.File, []igntypes.File) {
+	oldFiles := mapset.NewSetFromSlice(getFileNames(oldFilesConfig))
+	oldFilesMap := filesToMap(oldFilesConfig)
+	newFiles := mapset.NewSetFromSlice(getFileNames(newFilesConfig))
+	newFilesMap := filesToMap(newFilesConfig)
+
+	added := []igntypes.File{}
+	for file := range newFiles.Difference(oldFiles).Iter() {
+		added = append(added, newFilesMap[file.(string)])
+	}
+
+	removed := []igntypes.File{}
+	for file := range oldFiles.Difference(newFiles).Iter() {
+		removed = append(removed, oldFilesMap[file.(string)])
+	}
+
+	modified := []igntypes.File{}
+	for file := range newFiles.Intersect(oldFiles).Iter() {
+		candidate := newFilesMap[file.(string)]
+		// if !reflect.DeepEqual(newFile, oldFilesMap[changeCandidate.(string)]) {
+		// FIXME refactor this func out of daemon.go
+		if !checkFiles([]igntypes.File{candidate}) {
+			modified = append(modified, candidate)
+		}
+	}
+	return added, removed, modified
+}
+
+func calculateActions(oldIgnConfig, newIgnConfig igntypes.Config, diff *MachineConfigDiff) []actionResult {
+	if diff.osUpdate || diff.kargs || diff.fips || diff.kernelType {
+		return []actionResult{RebootPostAction{Reason: "OS/Kernel changed"}}
+	}
+
+	// Check for any changes not excluded by Reconcilable()
+	// Alternatively, fold this code into that function
+	if !reflect.DeepEqual(oldIgnConfig.Ignition, newIgnConfig.Ignition) {
+		return []actionResult{RebootPostAction{Reason: "Ignition changed"}}
+	}
+	if !reflect.DeepEqual(oldIgnConfig.Passwd, newIgnConfig.Passwd) {
+		return []actionResult{RebootPostAction{Reason: "Passwords changed"}}
+	}
+	if !reflect.DeepEqual(oldIgnConfig.Systemd, newIgnConfig.Systemd) {
+		return []actionResult{RebootPostAction{Reason: "Systemd changed"}}
+	}
+	if !reflect.DeepEqual(oldIgnConfig.Storage.Files, newIgnConfig.Storage.Files) {
+		actions := []actionResult{}
+
+		added, removed, modified := getFileChanges(oldIgnConfig.Storage.Files, newIgnConfig.Storage.Files)
+		if len(added) > 0 || len(removed) > 0 {
+			return []actionResult{RebootPostAction{Reason: fmt.Sprintf("Files were added or removed")}}
+		}
+		for _, file := range modified {
+			if file.Node.Path != "/etc/containers/registry.conf" {
+				return []actionResult{RebootPostAction{Reason: fmt.Sprintf("Storage file %v changed", file.Node.Path)}}
```

Comment:
s/Storage/Registy/g perhaps? Also, I think this must be abstracted out, we won't add if/else(s) to this

Reply:
I'm also in two minds about have flexible vs. simple this should be.
Definitely open to suggestions.

Reply:
At one point there was basically a lookup table that mapped filenames to actions, with reboot used for anything not present.
How would you feel about re-introducing that?